### PR TITLE
UnixPB: Update playbook to allow SLES12 to work

### PR DESCRIPTION
**Pull request by Willsparker**
fixes: #1033 

Fixes all of the issues listed in 1033. 

The only concern I have is the changes implemented to not install CPAN's XML::Parser, and instead install `perl-xml-parser` under the `test-tool-packages` section. Supposedly that's meant to be equivalent, according to this : https://github.com/chorny/XML-Parser/blob/master/README

However, I don't know how to verify this. It does install however.
I've done this as CPAN XML::Parser has a dependency for `libexpat1-dev` (or `libexpat-dev`) or `expat-dev` depending on the OS, that appears to not be available to SLES12-SP4 (The machine both @Haroon-Khel and I were testing on).

Note: Currently testing using the `vagrantPlaybookCheck.sh` script. https://ci.adoptopenjdk.net/view/Tooling/job/VagrantPlaybookCheck/220/

### CODE REVIEW

File: ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/NVidia_Cuda_Toolkit/tasks/main.yml [hunk 1]
```diff
@@ -104,6 +102,19 @@
     #TODO: rpm used in place of yum or rpm_key module
     - skip_ansible_lint
 
+- name: Sed change gpgcheck for SLES12 on x86_64
```
Comment (sxa555) [inline ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/NVidia_Cuda_Toolkit/tasks/main.yml]: Why is this needed - if we have to do this instead of doing a proper GPG check then it needs a comment explaining why

Comment (Willsparker) [reply ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/NVidia_Cuda_Toolkit/tasks/main.yml]: I'm not sure- It was part of the `Enable NVidia_Cuda_Toolkit for SLES x86_64` tasks, I just split it up as `rpm` was reading in every word as a package to install.

Comment (sxa555) [reply ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/NVidia_Cuda_Toolkit/tasks/main.yml]: Can you do some experimentation and see if it's still needed and if so what's the message - we should document it in this playbook

**Comment by Willsparker** [comment]
@sxa555 The sed task does appear to be necessary. When not running it the `install NVidia_Cuda_Toolkit` task outputs the following: 

`fatal: [9.20.195.225]: FAILED! => {"changed": false, "cmd": ["/usr/bin/zypper", "--quiet", "--non-interactive", "--xmlout", "refresh"], "msg": "Zypper run command failed with return code 4.", "rc": 4, "stderr": "", "stderr_lines": [], "stdout": "<?xml version='1.0'?>\n<stream>\n<prompt id=\"14\">\n<text>\nNew repository or package signing key received:\n\n&lt;gpgkey-info&gt;&lt;repository&gt;cuda&lt;/repository&gt;&lt;key-name&gt;cudatools &amp;lt;cudatools@nvidia.com&amp;gt;&lt;/key-name&gt;&lt;key-fingerprint&gt;AE09FE4BBD223A84B2CCFCE3F60F4B3D7FA2AF80&lt;/key-fingerprint&gt;&lt;key-created time_t=&quot;1466808197&quot; text=&quot;2016-06-24T15:43:17-07&quot;/&gt;&lt;key-expires time_t=&quot;0&quot; text=&quot;0&quot;/&gt;&lt;rpm-name&gt;gpg-pubkey-7fa2af80-576db785&lt;/rpm-name&gt;&lt;/gpgkey-info&gt;\n\nDo you want to reject the key, trust temporarily, or trust always?</text>\n<option default=\"1\" value=\"r\" desc=\"Don&apos;t trust the key.\"/>\n<option value=\"t\" desc=\"Trust the key temporarily.\"/>\n<option value=\"a\" desc=\"Trust the key and import it into trusted keyring.\"/>\n</prompt>\n<message type=\"error\">Repository &apos;cuda&apos; is invalid.\n[cuda|http://developer.download.nvidia.com/compute/cuda/repos/sles122/x86_64] Valid metadata not found at specified URL\nPlease check if the URIs defined for this repository are pointing to a valid repository.\n</message>\n<message type=\"error\">Skipping repository &apos;cuda&apos; because of the above error.</message>\n<message type=\"error\">Some of the repositories have not been refreshed because of an error.</message>\n</stream>\n", "stdout_lines": ["<?xml version='1.0'?>", "<stream>", "<prompt id=\"14\">", "<text>", "New repository or package signing key received:", "", "&lt;gpgkey-info&gt;&lt;repository&gt;cuda&lt;/repository&gt;&lt;key-name&gt;cudatools &amp;lt;cudatools@nvidia.com&amp;gt;&lt;/key-name&gt;&lt;key-fingerprint&gt;AE09FE4BBD223A84B2CCFCE3F60F4B3D7FA2AF80&lt;/key-fingerprint&gt;&lt;key-created time_t=&quot;1466808197&quot; text=&quot;2016-06-24T15:43:17-07&quot;/&gt;&lt;key-expires time_t=&quot;0&quot; text=&quot;0&quot;/&gt;&lt;rpm-name&gt;gpg-pubkey-7fa2af80-576db785&lt;/rpm-name&gt;&lt;/gpgkey-info&gt;", "", "Do you want to reject the key, trust temporarily, or trust always?</text>", "<option default=\"1\" value=\"r\" desc=\"Don&apos;t trust the key.\"/>", "<option value=\"t\" desc=\"Trust the key temporarily.\"/>", "<option value=\"a\" desc=\"Trust the key and import it into trusted keyring.\"/>", "</prompt>", "<message type=\"error\">Repository &apos;cuda&apos; is invalid.", "[cuda|http://developer.download.nvidia.com/compute/cuda/repos/sles122/x86_64] Valid metadata not found at specified URL", "Please check if the URIs defined for this repository are pointing to a valid repository.", "</message>", "<message type=\"error\">Skipping repository &apos;cuda&apos; because of the above error.</message>", "<message type=\"error\">Some of the repositories have not been refreshed because of an error.</message>", "</stream>"]}`

**Comment by Haroon-Khel** [comment]
Unfortunately the `perl-XML-parser` fails to install, with the following error:
```
    },
    "item": "perl-XML-parser",
    "msg": "Package '+perl-XML-parser' not found.",
    "rc": 104,
    "stderr": "",
    "stderr_lines": [],
    "stdout": "<?xml version='1.0'?>\n<stream>\n<message type=\"error\">Package &apos;+perl-XML-parser&apos; not found.</message>\n</stream>\n",
    "stdout_lines": [
        "<?xml version='1.0'?>",
        "<stream>",
        "<message type=\"error\">Package &apos;+perl-XML-parser&apos; not found.</message>",
        "</stream>"
    ]
}
```

**Comment by Willsparker** [comment]
> Unfortunately the `perl-XML-parser` fails to install, with the following error:

I saw that whilst testing the above NVidia_Cuda task. In that case, I'm not sure what to do regarding the `CPAN XML::Parser` task not working for SLES12. More detail: https://github.com/AdoptOpenJDK/openjdk-infrastructure/issues/1033#issuecomment-568457068

File: ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/CPAN/tasks/main.yml [hunk 2]
```diff
@@ -47,6 +47,7 @@
 
 - name: Install XML::Parser
   command: cpanm --with-recommends --sudo XML::Parser
+  when: ! (ansible_distribution == "SLES" and ansible_distribution_major_version == "12")
```
Comment (Haroon-Khel) [inline ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/CPAN/tasks/main.yml]: This doesnt skip, as it should, on SLES 12. It may be a syntax error.
Even if skipping this task on SLES 12 is a temp solution, I think the condition `ansible_distribution != "SLES" and ansible_distribution_major_version != "12"` would be more suitable

Comment (Willsparker) [reply ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/CPAN/tasks/main.yml]: Fix in 👍

**Comment by Willsparker** [comment]
With the latest commit, I've removed `perl-XML-parser` from `AdoptOpenJDK_Unix_Playbook/roles/Common/vars/SLES.yml` however an alternative is yet to be found.

**Comment by Haroon-Khel** [comment]
@Willsparker Turns out the playbook already installs `expat`, but only for SLES 11 (https://github.com/AdoptOpenJDK/openjdk-infrastructure/blob/master/ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Common/tasks/SLES.yml#L170) @sxa555 Would you happen to know if this is intentional or simply due to a failure to update the playbooks for SLES 12? 

Ive changed the condition to SLES 12. Consequently, the `Install XML::Parser` now gets further than before, but still fails. A dependency, `Time::Zone`, fails to install:
`FAIL Installing Time::Zone failed. See /root/.cpanm/work/1577975525.3460/build.log for details. Retry with --force to force install it.`

**Comment by Willsparker** [comment]
> Ive changed the condition to SLES 12. Consequently, the `Install XML::Parser` now gets further than before, but still fails. A dependency, `Time::Zone`, fails to install:
> `FAIL Installing Time::Zone failed. See /root/.cpanm/work/1577975525.3460/build.log for details. Retry with --force to force install it.`

Ah okay, thanks! I'll look to see if there's any way around that :-)
